### PR TITLE
decoder model uses pad left in generation task, the pad will happen i…

### PR DESCRIPTION
**Pull request by sywangyi**
…f batch > 1.

if the model is not optimized (static shape support), no need to pad right, or else it will pad check error like in https://huggingface.co/mosaicml/mpt-7b/blob/main/modeling_mpt.py#L156

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by sywangyi** [comment]
@yao-matrix

**Comment by sywangyi** [comment]
@regisss please help review

**Comment by sywangyi** [comment]
the output before the fix:

command:
python run_generation.py \
--model_name_or_path gpt2 \
--batch_size 2 \
--max_new_tokens 100 \
--use_hpu_graphs \
--use_kv_cache \
--bf16


----------------------------------------------------------------------

Input/outputs:
----------------------------------------------------------------------
input 1: DeepSpeed is a machine learning framework
output 1: DeepSpeed is a machine learning framework that can be used to predict the speed of a car.

The goal of the project is to create a machine learning framework that can be used to predict the speed of a car.

The project is a proof of concept.

The project is a proof of concept.

The project is a proof of concept.

The project is a proof of concept.

The project is a proof of concept.

The project is a proof of concept.


----------------------------------------------------------------------
input 2: He is working on
output 2: He is working on

a new version of the game, which will be released on the 3DS in Japan on May 18.

The game will be available on the 3DS in Japan on May 18.

The game will be available on the 3DS in Japan on May 18.

The game will be available on the 3DS in Japan on May 18.

The game will be available on the 3DS in Japan on May 18.

The game will be available on the
----------------------------------------------------------------------

after the fix:
Input/outputs:
----------------------------------------------------------------------
input 1: DeepSpeed is a machine learning framework
output 1: DeepSpeed is a machine learning framework that can be used to predict the speed of a car.

The goal of the project is to create a machine learning framework that can be used to predict the speed of a car.

The project is a proof of concept.

The project is a proof of concept.

The project is a proof of concept.

The project is a proof of concept.

The project is a proof of concept.

The project is a proof of concept.


----------------------------------------------------------------------
input 2: He is working on
output 2: He is working on a new book, "The Art of the Deal," which will be published by the publisher.

"I'm not sure if I can get a publisher to do it," he said. "I'm not sure if I can get a publisher to do it."

The book is a collection of essays by the author, who is a former editor of the New York Times bestseller list.

"I think it's a great book," he said. "I think it's
----------------------------------------------------------------------

**Comment by sywangyi** [comment]
@lvliang-intel

**Review by regisss — COMMENTED** [review]
Thanks for this PR @sywangyi! I left a couple of comments.

File: examples/text-generation/run_generation.py [hunk 1]
```diff
@@ -224,12 +225,12 @@ def main():
...
-        use_cache=args.use_kv_cache,
-        do_sample=args.do_sample,
-        num_beams=args.num_beams,
-    )
+    generation_config = {
+        "max_new_tokens": args.max_new_tokens,
+        "use_cache": args.use_kv_cache,
+        "do_sample": args.do_sample,
+        "num_beams": args.num_beams,
+    }
```
[the remark this answers is not in the document]
Comment (sywangyi) [reply examples/text-generation/run_generation.py:233]: there's the logic in https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/generation/utils.py#L284. the reason to do this is to left model generation config like "pad_id", eos id pass to generation.

File: examples/text-generation/run_generation.py [hunk 2]
```diff
@@ -283,8 +284,8 @@ def generate():
             outputs = model.generate(
                 **input_tokens,
                 **kwargs,
-                generation_config=generation_config,
-                lazy_mode=True,
+                **generation_config,
+                lazy_mode=args.use_hpu_graphs,
```
[the remark this answers is not in the document]
Comment (sywangyi) [reply examples/text-generation/run_generation.py:288]: > We should keep `lazy_mode=True` to let the possibility to users to perform generation in lazy mode without HPU graphs no?

just keep the same logic with https://github.com/huggingface/optimum-habana/blob/main/examples/text-generation/run_generation.py#L409

[the remark this answers is not in the document]
Comment (regisss) [reply examples/text-generation/run_generation.py:288]: Ah it seems I forgot to change this! Could you change line 409 with `lazy_mode=True` please?

File: examples/text-generation/run_generation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (regisss) [reply examples/text-generation/run_generation.py:233]: Hmm I don't see why using `GenerationConfig` prevents this

[the remark this answers is not in the document]
Comment (regisss) [reply examples/text-generation/run_generation.py:233]: `pad_token_id` and `eos_token_id` are attributes of `GenerationConfig`: https://github.com/huggingface/transformers/blob/04ab5605fbb4ef207b10bf2772d88c53fc242e83/src/transformers/generation/configuration_utils.py#L200

[the remark this answers is not in the document]
Comment (regisss) [reply examples/text-generation/run_generation.py:233]: I think we should rather have:
```
generation_config = copy.deepcopy(model.generation_config)
generation_config.max_new_tokens=args.max_new_tokens,
generation_config.use_cache=args.use_kv_cache,
generation_config.do_sample=args.do_sample,
generation_config.num_beams=args.num_beams,
```

[the remark this answers is not in the document]
Comment (regisss) [reply examples/text-generation/run_generation.py:233]: I'm a bit nitpicking here, but `GenerationConfig` gives us more possibility than a dictionary (mainly load/save a generation config from/to the hub)

[the remark this answers is not in the document]
Comment (sywangyi) [reply examples/text-generation/run_generation.py:233]: yes, you could explicitly define it in generation_config, but in the example, it has not done this. The logic of https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/generation/utils.py#L284 is if you do not pass GenerationConfig,  generation config will be initialized from the model generation_config firstly, then update it according to the kwargs. same with the logic you provide above

[the remark this answers is not in the document]
Comment (regisss) [reply examples/text-generation/run_generation.py:233]: Yes I agree. But setting the generation config explicitly in the example would enable saving it later.
But it's okay, let's keep it like this and we'll change it if we introduce a code block to send the generation config to the hub.

[the remark this answers is not in the document]
Comment (regisss) [reply examples/text-generation/run_generation.py:233]: Actually, with a dictionary, this warning is logged with the command line you gave above: https://github.com/huggingface/optimum-habana/blob/1eef41c53ff60a12f6466c2775df03cbc0b3ebea/optimum/habana/transformers/generation/utils.py#L290
So let's rather avoid it with a generation config as I proposed please.

[the remark this answers is not in the document]
Comment (regisss) [reply examples/text-generation/run_generation.py:233]: I know it's nitpicky but I would rather prevent the Transformers warning saying that it's a deprecated strategy. Some users would probably get confused.

[the remark this answers is not in the document]
Comment (sywangyi) [reply examples/text-generation/run_generation.py]: got it, update the PR

**Review by regisss — APPROVED** [review]
LGTM, thanks for addressing my comment @sywangyi !
Could you just run
```bash
pip install black ruff
make style
```
from the root of the repo please?

**Comment by regisss** [comment]
@sywangyi There are some merge conflicts since https://github.com/huggingface/optimum-habana/pull/281 just got merged.
It's basically about setting a few additional fields in the genration config.

**Comment by sywangyi** [comment]
> @sywangyi There are some merge conflicts since #281 just got merged. It's basically about setting a few additional fields in the generation config.

fix the conflict

**Review by regisss — APPROVED** [review]
LGTM!

**Pull request merged**